stl/inc/ranges

```diff
@@ -4749,6 +4752,477 @@ namespace ranges {
+            return _STD move(_Range);
+        }
+
+        _NODISCARD constexpr auto begin() {
+            constexpr bool _Use_const = _Simple_view<_Vw> && is_reference_v<_InnerRng<false>> && _Simple_view<_Pat>;
+            return _Iterator<_Use_const>{*this, _RANGES begin(_Range)};
+        }
+
+        _NODISCARD constexpr auto begin() const
+#ifdef __clang__ // TRANSITION, LLVM-47414
+            requires _Can_const_join_with<_Vw, _Pat>
+#else // ^^^ workaround / no workaround vvv
+            requires input_range<const _Vw> && forward_range<const _Pat> && is_reference_v<_InnerRng<true>>
+#endif // TRANSITION, LLVM-47414
+        {
+            return _Iterator<true>{*this, _RANGES begin(_Range)};
+        }
+
+        _NODISCARD constexpr auto end() {
+            constexpr bool _Both_simple = _Simple_view<_Vw> && _Simple_view<_Pat>;
+            if constexpr (forward_range<_Vw> //
+                          && is_reference_v<_InnerRng<false>> && forward_range<_InnerRng<false>> //
+                          && common_range<_Vw> && common_range<_InnerRng<false>>) {
+                return _Iterator<_Both_simple>{*this, _RANGES end(_Range)};
+            } else {
+                return _Sentinel<_Both_simple>{*this};
+            }
+        }
+
+        _NODISCARD constexpr auto end() const
+#ifdef __clang__ // TRANSITION, LLVM-47414
+            requires _Can_const_join_with<_Vw, _Pat>
+#else // ^^^ workaround / no workaround vvv
+            requires input_range<const _Vw> && forward_range<const _Pat> && is_reference_v<_InnerRng<true>>
+#endif // TRANSITION, LLVM-47414
+        {
+            if constexpr (forward_range<const _Vw> && forward_range<_InnerRng<true>> //
+                          && common_range<_Vw> && common_range<_InnerRng<true>>) {
+                return _Iterator<true>{*this, _RANGES end(_Range)};
+            } else {
+                return _Sentinel<true>{*this};
+            }
+        }
+    };
+
+    template <class _Rng, class _Pat>
+    join_with_view(_Rng&&, _Pat&&) -> join_with_view<views::all_t<_Rng>, views::all_t<_Pat>>;
+
+    template <input_range _Rng>
+    join_with_view(_Rng&&, range_value_t<range_reference_t<_Rng>>)
+        -> join_with_view<views::all_t<_Rng>, single_view<range_value_t<range_reference_t<_Rng>>>>;
+
+    namespace views {
+        struct _Join_with_fn {
+            // clang-format off
+            template <viewable_range _Rng, class _Pat>
+            _NODISCARD constexpr auto operator()(_Rng&& _Range, _Pat&& _Pattern) const noexcept(
+                noexcept(join_with_view(_STD forward<_Rng>(_Range), _STD forward<_Pat>(_Pattern)))) requires requires {
+                join_with_view(_STD forward<_Rng>(_Range), _STD forward<_Pat>(_Pattern));
+            }
+            { // clang-format on
+                return join_with_view(_STD forward<_Rng>(_Range), _STD forward<_Pat>(_Pattern));

```

Comment:
Can we use braces to construct the `join_with_view`? It appears that the existing `join_view` does so:
https://github.com/microsoft/STL/blob/27877181dc50fc5f0dc9d679703437eb105e2b9f/stl/inc/ranges#L3404-L3412

Reply:
I believe that we can't. We need to allow narrowing conversion here, and this matters e.g. if someone wants to join a sequence of strings with `std::toupper('x')`.